Add Transform and Grid based animations

smspillaz opened the pull request:
This branch adds some affine transformation animations, grid transformation animations and associated helpers. (Note: This work was started about a year ago while I was still at Endless, but was never actually finished and I didn't get much of a chance to work on it due to my study commitments, but I've finally had the time to refactor things into what is a hopefully better state now. I said I would eventually get around to this, so here it is :-)).

Stepping the animations is done using Stepper implementations (just a function object at the moment, an interface in the GObject interface). They can be composed as functions of each other, for instance the Reverse stepper is a composition of 1.0f - baseStepper(ms).

All the transform animations themselves are implementations of the TransformAnimation interface - the TransformAnimation class does not have any concrete implementation details. The animations themselves share code through the added header files with utility methods.

The actual matrix math is implemented using glm, a popular C++ header only library used for doing matrix transformations.

The GridAnimation template and helper functions, can be used to animate surfaces by deforming equally spaced vertices on a texture.

Changed from #4 and #8 is that the C++ animations now no longer require all properties to be specified in order to be constructed. Instead, they each have a default constructor, which sets sane defaults for the parameters of the animation. I also added a "property" system which allows callers to get/set properties on the animation in an ABI-safe way without too much boilerplate. This change (combined with a few other things), allows us to then drop a bunch of constructor abuse from the GObject wrappers - we now just construct the animation using the default constructor and let the property system handle the rest (though there are a few exceptions for object-valued properties, which comes from the fact that it doesn't make sense to allow those properties to be `NULL` and the defaults set on the C++ side are not bindings-friendly implementations).

Also changed from #4 and #8: 
 - The steppers are no longer C++ closures. This didn't work too well with bindings. Instead they are objects, with virtual methods (eg, the `ReverseStepper` takes a `Stepper` subclass as its argument, where we can pass a `glib::StepperWrapper` in order to wrap the GObject stepper implementation directly, as opposed to having to create all these intermediary GObjects every time a property is accessed.
 - Introduced `animation::BoxQuery` - which is explained a little more in the relevant commit, but it is essentially a mechanism to keep track of changes to surface geometry during the animation.
 - Added tests on the GJS side to make sure that the bindings work correctly
 - Moved the clutter integration into libanimation itself (meaning that less code needs to be carried in the shell)
 - Created a gnome-shell extension so this can be tested directly.

https://phabricator.endlessm.com/T23543
https://phabricator.endlessm.com/T23697

Commit:
build: Add header-dependency on glm

We're going to do affine transformations and glm is a particularly
well respected library for things like that. Note that the dependency
is a submodule, so you'll need to fetch and pull submodules before
building.

Commit:
animation: Add box type

The "box" type represents a simple perpendicular quadrilateral
in 2D space given by its top left and bottom right co-ordinates
(eg, x1, y1, x2, y2).

Commit:
animation: Add a TransformAnimation base class

We'll do some affine transformations based on this one.

There's three main methods to override here:

 - Step() -> Take a single step in the animation by a given
             number of milliseconds. Returns true if more
             steps need to be taken, false otherwise
 - Matrix() -> Get the transformation matrix for the animated
               surface at this point in the animation.
 - Extremes() -> Get the four points specifying the 3D bounding
                 volume for this surface, usually transforming
                 the co-ordinates of the surface in scene-relative
                 co-ordinates that have been passed in.

Commit:
animation: Add zoom animation

This is a simple affine transformation that goes from
one box to another.

The animation runs "from" some source point "to" some target point
which is the surface' "natural" position within the scene (eg, if
there were a scene graph, the "to" position forms its co-ordinates
and its geometry).

For instance, if the window was minimizing, then the animation would
have to be run in reverse, "from" the minimized point "to" the unminimized
point, but stepping backwards.

Commit:
animation: Add bounce animation

This animation simulates a gentle bounce on a sine wave from
the center of the window outwards.

The best way to think of this animation is to think of an attenuating
sine wave squeezed inwards by two bounds that converge on a single point.

In thise case those lines are running from the "initialScale" to 1.0
and the "maximumScale" down to 1.0. We run the sine wave for 2pi * nBounce
iterations (scaling it it to fit within the time range), but the effect
of the naimation is scaled according to where we are on the bunds.

Now, rotate the attenuating sine wave so that it is facing towards you
on the z-axis (OpenGL co-ordinates). This is essentially what the animation
is.

Commit:
animation-glib: Add constructor helpers

These helpers are used to quickly append a new "interface"
construction prop to the existing construction properties
and make it easy to lookup the props inline.

Basically, we need these helpers because we are wrapping some
C++ interface implementation in a GObject. GObject doesn't require
all the properties to be set at construction time, whereas
constructing the C++ interface implementation might. So we need to
collect all the properties in the constructor and then construct
the underlying C++ interface there, so that it is immediately
available to the base class as a property on the object.

Doing this by looking over all the construct properties and assigning
values would be cumbersome, especially doing it for every effect
implementation. Instead, we put all the construct properties into
a hash table (hashed by the property name) and then fetch them in
O(1) using the InterfaceConstructor template.

This template constructs the "Interface" type (the type passed
should be the implementation of the interface you want to construct)
and variadically takes in all the construction parameters. Use
ForwardValueFromHT to extract a typed value from the given property name
from its corresponding GValue in the construct properties hash table. For
instance, supposing that MyType's constructor is defined as follows:

MyType::MyType (float                   bar,
                animation::Point const &baz)
{
  ...
}

You would use the following in the constructor override for the GObject
wrapper:

const char *interesting_props = {
  "bar",
  "baz",
  NULL
};
g_autoptr(GHashTable) props_ht =
  static_hash_table_of_values_for_specs (interesting_props,
                                         construct_params,
                                         n_construct_params);
auto *interface = InterfaceConstructor <MyType>::construct (
  ForwardValueFromHT (props_ht, g_value_get_float, "bar"),
  ForwardValueFromHT (props_ht, animation_point_from_gvalue, "baz")
);
g_auto(GValue) interface_value = G_VALUE_INIT;
unsigned int new_n_construct_params = 0;
GObjectConstructParam *new_construct_params =
  append_interface_prop_to_construct_params (construct_params,
                                             n_construct_params,
                                             object_class,
                                             &interface_value,
                                             interface,
                                             &new_n_construct_params);

object_class->constructor (type,
                           new_construct_params,
                           new_n_construct_params);

Another thing you may wish to do is ensure that certain properties
always take a certain value that is not their default (for instance,
you might want to ensure tha a certain object-valued property
is some non-NULL sane default in the event that it is not set). To
do that, you can use replace_named_pointer_props_in_construct_params_if_null,
which will only construct the desired placeholder object property values
in the event that the property provided by the user or by default
due to the G_PARAM_SPEC_CONSTRUCT flag was NULL. For example:

  /* We need to also set defaults for certain properties in order
   * to ensure that they are bindable later on, in case they are
   * not set by the caller. */
  replace_named_pointer_props_in_construct_params_if_null (
    construct_params,
    n_construct_params,
    {
      ReplacePropSpec ("stepper", g_value_get_object, g_value_set_object, []() -> gpointer {
        return animation_linear_stepper_new (1);
      }),
      ReplacePropSpec ("target", g_value_get_object, g_value_set_object, []() -> gpointer {
        return animation_box_query_new ();
      })
    }
  );

Note that with most of the animations in this library, you can get away with
the latter only - which is to say that most of the animations can be default-constructed,
however their default properties may not be easily bindable because they refer to some
object that is not being wrapped by a GObject that can be exposed to the introspection
system. Therefore, the animation author's job is to ensure that the properties
set on the underlying C++ type support bindability.

Commit:
tests: Add glm_ostream_operators.h

This will be useful for printing out the contents of
matrices when tests fail. Also necessary if we want to use
comparison matchers with them.

Commit:
animation: Add box_calculation inline functions

We'll use these later for the affine transformation animations
to calculate box offsets.

Usually we want to compute where the center of a box is in both
absolute co-ordinates and relative co-ordinates. ComputeBoxCenter
will do it in absolute co-ordinates and ComputeBoxCenterOffset
in co-ordinates relative to the top left corner of the box.

Commit:
animation: Add math inline helper functions

We'll use these later when implementing linear progress based
animations.

Right now we only have clamp(), used to ensure that the given value
sits between some bounds.

Commit:
animation: Add 4D vector type

Right now we only have 2D vector types, but some consumers like
clutter want bounding volumes in a four dimensional space, so we
need a type to represent such bounding volumes.

Commit:
matchers: Add AlmostEqMatcher

This matcher allows the verify that one type instance is
"almost equal" to another type instance. The way that is
done is that the type must implement the
animation::matcher::AlmostEqTrait type trait and define
the apply() method.

Commit:
grid: Add GridAnimation interface

The GridAnimation interface represents transformations that are done
on a grid of vertices  that a surface is subdivided into. The
transformation can be any arbitrary function. Implementations must
override three functions:

 - Step() -> Take a single step on the function according to the
             number of milliseconds passed. If more steps are needed
             to complete the animation, return true, otherwise return
             false.
 - DeformUVToModelSpace() - Given some co-ordinates between (0, 0) and
                            (1, 1), deform the co-ordinates into model
                            space, that is the space in the scene of
                            the surface as if the function were applied.
 - Extremes() - Get the geometry of a 2D bounding plane totally encompassing
                the deformed surface.

Commit:
animation: Add magic lamp animation

This animation simulates the "genie" effect on another popular
operating system, appearing to "suck" the window into a destination
rectangle.

The way it works is a little complicated. Basically, we subdivide
the surface into a very high resolution grid and from each point
in the source geometry to each point in the target geometry, we
we have a sigmoid function (S-curve, eg e^x / (1 + ke^px) that each
grid point travels down towards its destination.

When deforming a UV into the pre-computed grid we must do a linear
interpolation along both axes, finding the nearest subdivision source
point on a UV mesh, then interpolating into that mesh component
on the scene-relative internal mesh.

https://phabricator.endlessm.com/T23697

Commit:
tests: Add ostream operator for Vector4D

This is necessary in order to use it with some of the testing
operators (like agd::Equals or agd::AlmostEq), since those
testing operators use operator<< in order to print out the
contents of the type in the event of a matching failure.

Commit:
geometry: Add operator== for point types

If we know that two points are going to be of the same type,
it is convenient to be able to compare them with the == operator
as opposed to having to use agd::equals

Commit:
animation: Add glide animation

This animation rotates the surface from some rotated position
towards a resting position where its rotation angle is
0, 0, 0 on all three axes. The axis that the surface can rotate
on is configurable in unit-cordinates. Rotation on 0.5, 0.5 rotates
at the center of the surface, whereas rotating from 0, 0 would rotate
from the top left corner.

Commit:
animation: Add property.h header

These are just some simple macros to define ABI safe property
get/set functions. They are accessed by calling the correct overload,
for instance, to set properties:

    animation.Target(target).Source(source)

And to get properties:

    auto const &target = animation.Target();

Commit:
animation: Put animation toplevel headers in variable

Then install them. This extends on #9, just adds some consistency in the
build definition.

smspillaz committed:
glib: Save Animation GirTarget in variable

smspillaz committed:
build: Build c and cpp

smspillaz committed:
animation: Add animation::BoxQuery interface

When animating from one position to another, it is possible that
some things could change underneath us. For instance, the target position
of the surface may be changed by the user while the animation is in progress.

In this case, the animation needs to adapt to this change by allowing
the caller some mechanism to transparently change the target box
tha the surface is animating to. This can be done by overriding
the animation::BoxQuery class and calling the Update() method as
appropriate, which will update the saved geometry of the target position.

For example, in a windowing system, one might set up a listener
for surface geometry events in their animation::BoxQuery subclass,
which then calls Update() when the geometry has been changed. This
obviates the need for the caller to explicitly update all animations.

smspillaz committed:
animation-glib/vector: Fix headerblock description

smspillaz committed:
animation: Add necessary includes to animation/geometry.h

smspillaz committed:
tests: Add JS tests and common JS tests utils

These include matrixMultiplyVector (multiplying a 4D matrix with a
4D vector), BoxWrapper (wraps a Animation.Box in an Animation.BoxQuery),
and applyCallerTranslation

smspillaz committed:
tests: Add meson build file for JS tests, with testStepper.js as the first

smspillaz committed:
animation-clutter: Add animation-clutter library and common helpers

This library will be used to bind libanimation to Clutter, so that
it can be used in GNOME-Shell. animation-clutter-common is a
set of helper functions that are used to translate boxes into the
correct space (eg, expanding paint boxes to cover the FBO size
for a deform effect, determining the paint box extents of an
actor, etc).

smspillaz committed:
animation-clutter: Add AnimationClutterActorBoxQuery

This class extends AnimationBoxQuery by attaching to an actor
and listening for changes in the actor geometry, updating the
underlying stored geometry in the box as appropriate. The caller
can then query the geometry directly from the box as opposed to
having to reach all the way into Clutter to get the geometry of
the animated surface.

smspillaz committed:
animation-clutter: Add AnimationClutterAffineEffect

This wraps any AnimationTransformAnimation subclass in a
ClutterEffect and uses the underlying AnimationTransformAnimation
to apply some sort of linear transformation to the actor
when it is painted depending on the state of the animation.

smspillaz committed:
animation-clutter: Add AnimationClutterGridEffect

AnimationClutterGridEffect extends ClutterDeformEffect and uses the
underlying AnimationGridAnimation subclass to deform the vertices of the actor
based on the state of the underlying animation.

smspillaz committed:
animation-clutter: Add AnimationClutterWobblyEffect

This extends ClutterDeformExtent and wraps AnimationWobblyModel
to deform the actor geometry in accordance with the underlying model
state.

smspillaz committed:
debian: Add build-dependency on libmutter-4-dev

smspillaz committed:
debian: Move gir1.2-animation-glib to libs from non-free/libs

smspillaz committed:
debian: Add debian packages for libanimation-clutter0

smspillaz committed:
compositor-plugins/gs: Add GNOME-Shell extension for libanimation

This uses AnimationClutter to provide animations for GNOME-Shell. It
listens for certain windows events, then runs animations according
to the settings in the org.gnome.shell.extensions.animation schema.

smspillaz committed:
debian: Add debian packaging for gnome-shell-extension-animation

smspillaz committed:
meson: Bump meson version to 0.42.0

eos-shell-build commented:
I don't recognize the submitter as a member of @endlessm. Can somebody in @endlessm verify the patch? See https://phabricator.endlessm.com/w/software/build/ci-integration/#pull-requests for more details.

smspillaz committed:
animation-glib: Use files() in meson.build

smspillaz committed:
animation: Add concept of a "stepper"

Steppers encapsulate some progress change within the animation. The
typical case is linear stepping, which just interpolates from
0 to 1. However, another case is reverse stepping which is
composed by 1.0f - linear (ms).

Future steppers can be added to add all sorts of easing parameters.

https://phabricator.endlessm.com/T23543

smspillaz committed:
animation: Use files([]) for list-of-files specification in meson.build

smspillaz committed:
travis: Add libmutter-4-dev to .travis.yml

smspillaz committed:
travis: Upgrade docker image to cosmic

We need this for libmutter-4-dev

smspillaz committed:
travis: Add repository for jasmine-gjs and install it

smspillaz commented:
Note: Travis is failing. I had a look into it, its some complications that come with trying to get the build dependencies in the docker container. I'm sure its fixable, may just time some more iterations...

smspillaz committed:
travis: Disable sanitizers

Seems like newer ubuntu images complain about the way that meson sets
these up and linking with asan seems to hang, so we need to disable it for now.

smspillaz committed:
travis: Use meson test as opposed to test binary directly

We now have additional JS tests that we would like to run

smspillaz committed:
tests: Fix headerblock file path comment

smspillaz committed:
travis: Just build jasmine-gjs instead of using Ubuntu package

The built package on OBS is out of date and not compatible with
mozjs60

smspillaz commented:
Okay, travis is passing now.

Commit:
build: Use Clutter-7, now

smspillaz committed:
animation-clutter: get_paint_volume -> modify_paint_volume

smspillaz committed:
affine-effect: clutter_actor_continue_paint needs a ClutterPaintContext

smspillaz committed:
animation-clutter: ClutterVertex -> graphene_point3d_t

smspillaz committed:
tests: Add missing stepper test

smspillaz committed:
shell: Drop use of Tweener

It isn't used in newer shell versions. Hurrah!

smspillaz committed:
shell: Don't disconnect signal on string

smspillaz committed:
animtion-clutter: Upgrade to Clutter-8

smspillaz committed:
animation-clutter: Use graphene_matrix_t

smspillaz committed:
wobbly_internal: No need for long double

smspillaz committed:
wobbly_internal: Save some computation

smspillaz committed:
compositor-plugins/gnome-shell: Connect to position-changed

smspillaz committed:
animation-clutter: No need to apply box offsets on wobbly effect anymore

We can just query the extremes and use those

smspillaz committed:
wobbly: some attempts to optimize things
